src/main/java/com/rabbitmq/http/client/ReactorNettyClient.java

```diff
@@ -511,98 +503,101 @@ public static String basicAuthentication(String username, String password) {
-            .onErrorMap(this.errorHandler)
-            .transform(decode(type));
+        return client.headers(authorizedHeader())
+            .doOnResponse(onResponseCallback)
+            .get()
+            .uri(uri(pathSegments))
+            .responseSingle(decode(type));
+    }
+
+    protected <T> BiFunction<? super HttpClientResponse, ? super ByteBufMono, ? extends Mono<T>> decode(Class<T> type) {
+        return (response, byteBufFlux) -> {
+            if (response.status().code() == 404) {
+                return Mono.empty();
+            } else {
+                return byteBufFlux.asByteArray().map(bytes -> {
+                    try {
+                        return objectMapper.readValue(bytes, type);
+                    } catch (IOException e) {
+                        throw Exceptions.propagate(e);
+                    }
+                });
+            }
+        };
     }
 
     @SuppressWarnings("unchecked")
     private <T> Flux<T> doGetFlux(Class<T> type, String... pathSegments) {
         return (Flux<T>) doGetMono(Array.newInstance(type, 0).getClass(), pathSegments).flatMapMany(items -> Flux.fromArray((Object[]) items));
     }
 
+    protected Consumer<HttpHeaders> authorizedHeader() {
+        return headers -> {
+            // FIXME try to get the token as late as possible?
+            headers.set(HttpHeaderNames.AUTHORIZATION, token.block());
+        };
+    }
+
     private Mono<HttpResponse> doPost(Object body, String... pathSegments) {
-        return client.post(uri(pathSegments), request -> Mono.just(request)
-            .transform(this::addAuthorization)
-            .map(ReactorNettyClient::disableChunkTransfer)
-            .map(ReactorNettyClient::disableFailOnError)
-            .transform(encode(body)))
+        return client.headers(JSON_HEADER.andThen(authorizedHeader()))
+            .noChunkedTransfer()
+            .doOnResponse(onResponseCallback)
+            .post()
+            .uri(uri(pathSegments))
+            .send(bodyPublisher(body))
+            .response()
             .map(ReactorNettyClient::toHttpResponse);
     }
 
     private Mono<HttpResponse> doPut(Object body, String... pathSegments) {
-        return client.put(uri(pathSegments), request -> Mono.just(request)
-            .transform(this::addAuthorization)
-            .map(ReactorNettyClient::disableChunkTransfer)
-            .map(ReactorNettyClient::disableFailOnError)
-            .transform(encode(body)))
+        return client.headers(JSON_HEADER.andThen(authorizedHeader()))
+            .noChunkedTransfer()
+            .doOnResponse(onResponseCallback)
+            .put()
+            .uri(uri(pathSegments))
+            .send(bodyPublisher(body))
+            .response()
             .map(ReactorNettyClient::toHttpResponse);
     }
 
+    private Mono<ByteBuf> bodyPublisher(Object body) {
+        return Mono.fromCallable(() -> {
+            // FIXME is the ByteBuf released by Reactor?
```

Comment:
The bytebuf is always released once by netty outbound writer. You could be using PooledByteBufAllocator default singleton (like Spring) or the one configured in the HttpClient builder if any.

Reply:
OK, thanks for the tips.

Reply:
@smaldini I can't find any `ByteBufAllocator` in the `HttpClient` or even a builder. Are you refering to the Reactor Netty 0.7?